rust/bridge/node/futures/src/executor.rs

```diff
@@ -0,0 +1,110 @@
+//
+// Copyright 2021 Signal Messenger, LLC.
+// SPDX-License-Identifier: AGPL-3.0-only
+//
+
+use neon::prelude::*;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Poll, Wake};
+
+/// Adds support for executing futures on a Neon [EventQueue][].
+///
+/// [EventQueue]: https://docs.rs/neon/0.7.0-napi.3/neon/event/struct.EventQueue.html
+pub trait EventQueueEx {
+    /// Schedules the future to run on the JavaScript main thread until complete.
+    fn send_future(&self, future: impl Future<Output = ()> + 'static + Send);
+}
+
+impl EventQueueEx for EventQueue {
+    fn send_future(&self, future: impl Future<Output = ()> + 'static + Send) {
+        self.send(move |mut cx| {
+            let task = Arc::new(FutureTask {
+                queue: cx.queue(),
+                future: Mutex::new(Some(Box::pin(future))),
+            });
+            task.poll();
+            Ok(())
+        })
+    }
+}
+
+/// Used to "send" a task from a thread to itself through a multi-threaded interface.
+struct AssertSendSafe<T>(T);
+unsafe impl<T> Send for AssertSendSafe<T> {}
+
+impl<T: Future> Future for AssertSendSafe<T> {
+    type Output = T::Output;
+    fn poll(self: Pin<&mut Self>, context: &mut std::task::Context) -> Poll<T::Output> {
+        // See https://doc.rust-lang.org/std/pin/index.html#projections-and-structural-pinning
+        let future = unsafe { self.map_unchecked_mut(|s| &mut s.0) };
+        future.poll(context)
+    }
+}
+
+/// Adds support for executing closures and futures on the JavaScript main thread's microtask queue.
+pub trait ContextEx<'a>: Context<'a> {
+    /// Schedules `f` to run on the microtask queue.
+    ///
+    /// Equivalent to `cx.queue().send(f)` except that `f` doesn't need to be `Send`.
+    fn run_on_queue(&mut self, f: impl FnOnce(TaskContext<'_>) -> NeonResult<()> + 'static) {
```

Comment:
It's public API, yeah. I don't think that needs to be commented since it's already a `pub trait`. It seemed weird to support this for futures and not for plain old tasks. (It's also API I was showing off to the Neon folks, above.)